gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ExecuteGobblinWorkflowImpl.java

```diff
@@ -123,4 +171,47 @@ protected static WUProcessingSpec createProcessingSpec(Properties jobProps, Even
     }
     return wuSpec;
   }
+
+  private void cleanupWorkDirs(WUProcessingSpec workSpec, EventSubmitterContext eventSubmitterContext, Set<String> directoriesToClean)
+      throws IOException {
+    // TODO: Add configuration to support cleaning up historical work dirs from same job name
+    FileSystem fs = Help.loadFileSystem(workSpec);
+    JobState jobState = Help.loadJobState(workSpec, fs);
+    // TODO: Avoid cleaning up if work is being checkpointed e.g. midway of a commit for EXACTLY_ONCE
+
+    if (PropertiesUtils.getPropAsBoolean(jobState.getProperties(), ConfigurationKeys.CLEANUP_STAGING_DATA_BY_INITIALIZER, "false")) {
+      log.info("Skipping cleanup of work dirs for job due to initializer handling the cleanup");
+      return;
+    }
+
+    try {
+      DirDeletionResult dirDeletionResult = deleteWorkDirsActivityStub.delete(workSpec, eventSubmitterContext,
+          calculateWorkDirsToDelete(jobState.getJobId(), directoriesToClean));
+      for (String dir : directoriesToClean) {
+        if (!dirDeletionResult.getSuccessesByDirPath().get(dir)) {
+          log.error("Directory {} was not cleaned up, please clean up manually", dir);
+        }
+      }
```

Comment:
you've got the exception handling above in the `finally`, but this then needs to throw an exception for that:
```
List<String> failures = directoriesToClean.stream().filter(dir -> !dirDeletionResult.getSuccessesByDirPath().get(dir)).collect(Collectors.toList);
if (failures) {
  log.error(...)
  throw new IOException(...)
}
```